packages/Discretization/src/DTK_PointSearch_def.hpp

```diff
@@ -0,0 +1,736 @@
+/****************************************************************************
+ * Copyright (c) 2012-2017 by the DataTransferKit authors                   *
+ * All rights reserved.                                                     *
+ *                                                                          *
+ * This file is part of the DataTransferKit library. DataTransferKit is     *
+ * distributed under a BSD 3-clause license. For the licensing terms see    *
+ * the LICENSE file in the top-level directory.                             *
+ ****************************************************************************/
+
+#ifndef DTK_POINT_SEARCH_DEF_HPP
+#define DTK_POINT_SEARCH_DEF_HPP
+
+#include <DTK_DBC.hpp>
+#include <DTK_DetailsTeuchosSerializationTraits.hpp>
+#include <DTK_DetailsUtils.hpp>
+#include <DTK_DistributedSearchTree.hpp>
+#include <DTK_PointInCell.hpp>
+#include <DTK_Topology.hpp>
+
+namespace DataTransferKit
+{
+namespace internal
+{
+template <typename DeviceType>
+KOKKOS_FUNCTION void computeBlockCellsBoundingBox(
+    unsigned int const dim, int const i, unsigned int const n_nodes,
+    unsigned int const node_offset, unsigned int const topo_id,
+    Kokkos::View<unsigned int *, DeviceType> cells,
+    Kokkos::View<unsigned int *, DeviceType> offset,
+    Kokkos::View<double **, DeviceType> coordinates,
+    Kokkos::View<double ***, DeviceType> block_cells,
+    Kokkos::View<Box *, DeviceType> bounding_boxes,
+    Kokkos::View<unsigned int **, DeviceType> bounding_box_to_cell )
+{
+    Box bounding_box;
+    // If dim == 2, we need to set bounding_box.minCorner()[2] and
+    // bounding_box.maxCorner[2].
+    if ( dim == 2 )
+    {
+        bounding_box.minCorner()[2] = 0;
+        bounding_box.maxCorner()[2] = 1;
+    }
+    unsigned int const k = offset( i );
+    for ( unsigned int node = 0; node < n_nodes; ++node )
+    {
+        unsigned int const n = node_offset + node;
+        for ( unsigned int d = 0; d < dim; ++d )
+        {
+            // Copy the coordinated in block_cells
+            block_cells( k, node, d ) = coordinates( cells( n ), d );
+            // Build the bounding box.
+            if ( block_cells( k, node, d ) < bounding_box.minCorner()[d] )
+                bounding_box.minCorner()[d] = block_cells( k, node, d );
+            if ( block_cells( k, node, d ) > bounding_box.maxCorner()[d] )
+                bounding_box.maxCorner()[d] = block_cells( k, node, d );
+        }
+    }
+    bounding_boxes( i ) = bounding_box;
+    bounding_box_to_cell( i, topo_id ) = k;
```

Comment:
No that's not what it does. The search works on bounding boxes which is a flat array but `PointInCell` works on block_cells that are multi-dim arrays and sorted by topologies. This creates a map so that I know from the output of search (bounding_boxes) which are the candidates (block_cells) to send to `PointInCell`.